### ISSUE TITLE
'event_del()' not always returns after the last event callback termination

### ISSUE THREAD

**Issue by jmillan** [libevent/libevent#225]
Hi,

I'm using release-2.0.21-stable version on Linux CentOS 6 and 7.

I've been using 'event_del()' as a mechanism to protect data which is potentially accessible by the main thread and a second one.

Everything worked as expected thanks to the current event waiting mechanism implemented in 'event_del_internal()', until I made some stress tests where incoming network traffic rate is higher than the event callback processing rate, generating congestion. In this case, 'event_del()' waits until the current event callback finishes but the callback is fired again for another incoming packet just before 'event_del()' returns, making the 'protected data' accessible by the two threads.

I put some debug lines right before and after 'event_del()' and right before and after the event callback function.

`Note: the event being deleted is that associated with fd 811`

```
[#7f7edeb06700] [callback, AmRtpSocket.cpp:189] INFO:  callback
[#7f7ee142f700] [removeFromReceiver, AmRtpSocket.cpp:175] INFO:  event_del
[#7f7edeb06700] [callback, AmRtpSocket.cpp:213] INFO:  callback finished
[#7f7edeb06700] [log_libevent_cb, log.cpp:87] INFO:  libevent debug: [epoll_dispatch: epoll_wait reports 3]
[#7f7edeb06700] [log_libevent_cb, log.cpp:87] INFO:  libevent debug: [event_active: 0x7f7e9004c4a0 (fd 811), res 2, callback 0x62bdd0]
[#7f7edeb06700] [log_libevent_cb, log.cpp:87] INFO:  libevent debug: [event_active: 0x1efea50 (fd 54), res 2, callback 0x7f7ee4270e82]
[#7f7edeb06700] [log_libevent_cb, log.cpp:87] INFO:  libevent debug: [event_active: 0x7f7eb80046b0 (fd 795), res 2, callback 0x62bdd0]
[#7f7edeb06700] [log_libevent_cb, log.cpp:87] INFO:  libevent debug: [event_process_active: event: 0x7f7e9004c4a0, EV_READ  call 0x62bdd0]
[#7f7edeb06700] [callback, AmRtpSocket.cpp:189] INFO:  callback
[#7f7ee142f700] [removeFromReceiver, AmRtpSocket.cpp:177] INFO:  event_del finished
```

It can be seen that the callback is running at the time 'event_del()' returns, exposing the 'protected data' to both threads and therefore making the access to it non secure.

The issue (if this behaviour isn't intentional) resides in the fact that 'event_del_internal()' blocks before actually deleting the event from the queue. During that blocking time new network packets can arrive making the corresponding callback being fired by the main thread just after the current one finishes.

The solution I've adopted is deleting the event from the queue before blocking in 'event_del_internal()'. This way I can assure that the current event callback is in fact the last event callback for that event.

I'll make a PR in case this is considered worthy of inclusion in the library.

I apologise in advance if the current is the desired behaviour, in which case I'd like to ask for a mechanism to protect the data handled by the event callback from being accessed by a second thread.

Thanks a lot.

**Comment by jmillan** [comment]
Let me please make myself clear.

The  commit 6b4b77a26590f states:
`Do not allow event_del(ev) to return while that event's callback is executing in another thread.  This fixes a nasty race condition.`

That statement is true if the `event's callback` there refers to the `event's callback current execution`.

Actually event_del(ev) can return while that event's callback is executing in another thread; not the same execution that was being running when entering event_del(ev) but a new one.

This is how it can happen, as shown in the logs provided in previous comment:

Thread 'A' is the main thread running the event's callbacks.
Thread 'B' is the process executing event_del(ev).

- 'B' executes event_del(ev)
- 'B' waits for the current callback to finish
- Before the current callback finishes a network packet arrives to the socket descriptor associated with the event
- 'A' finishes the current callback execution
- The event gets active again due to the incoming packet and the callback is executed again in 'A'
- 'B' gets unlocked due to the previous callback execution termination, removes the event from the queue and returns from event_del(ev)


The solution I propose (attached in the pool request) is removing the event form the queue `before` waiting for the event's callback current execution termination, and no after. This way it is avoided that the event gets active again just after the current event's callback execution finishes.

The call flow changes to:

- 'B' executes event_del(ev)
- 'B' deletes the event from the queue
- 'B' waits for the current callback to finish
- Meantime a network packet arrives to the socket descriptor associated with the event
- 'A' finishes the current callback execution
- The event doesn't get active since it is deleted from the queue
- 'B' gets unlocked due to the previous callback execution termination and returns from event_del(ev)


Any comment regarding to this issue is very appreciated

**Comment by ellzey** [comment]
This __may__ be fixed in 2.1, but I need to come up with some test code to re-produce the original issue in 2.0-stable. Will get back to you soonish.

**Comment by azat** [comment]
@jmillan why after first event callback execution finished, `event_del()` didn't returned?

Since this is how it suppose to work. Of course you can change event's base with `event_base_set()` and in this case you can even move event to another thread, but you can do this only for non-active/pending events so this is not your case.

**Comment by azat** [comment]
@jmillan friendly ping

**Comment by jmillan** [comment]
Hi @azat,

 > @jmillan why after first event callback execution finished, event_del() didn't returned?

Correct.

If you take a look at my first comment: 


> 'event_del()' waits until the current event callback finishes but the callback is fired again for another incoming packet just before 'event_del()' returns, making the 'protected data' accessible by the two threads.

> I put some debug lines right before and after 'event_del()' and right before and after the event callback function.


Also, the logs provided in the first comment  indicate that a callback is being executed when calling `event_del`. This function waits for the `current callback` to finish and then `another` event fires the callback again. `event_del` function returns while the second callback is running.

Kind regards

**Comment by azat** [comment]
Okay, I think that we do need this patch (#236) at least to fix the case
when event_del() can return while that event's callback is executing in
another thread. (@ellzey agree?)

@jmillan can you write a regression unit test for this (you can take a
look at referenced test in this issue)?
That would help a lot!

And many thanks for pointing out to details again!

**Comment by jmillan** [comment]
> @jmillan can you write a regression unit test for this (

Yes, give me some time please, I'll do it as soon as I can.

**Comment by jmillan** [comment]
Hi,

I don't know how to accomplish the regression test. Any help is very welcome.

The issue is reproduced for the following scenario, considering the next actions happen in chronological order.

```
Main event thread: event callback start
Another thread:    event_del_start
Main event thread: event callback end
Main event thread: event callback start
Another thread:    event_del_end
Main event thread: event callback end   
```

**Comment by PugalendranDell** [comment]
@azat @nmathewson @provos - I am seeing the same problem as discussed in this mail thread in latest debian stretch. 

Latest debian stretch uses 2.0.21-stable libevent. Do you guys have any plan to patch this change to libevent2.0.x ? In my code base, moving to latest libevent 2.1.x is not possible now, do you guys have any other way to get these changes in latest stretch debian ?

**Comment by azat** [comment]
>I am seeing the same problem as discussed in this mail thread in latest debian stretch.

Can you provide **minimal** reproducer (100 lines of code or similar)?

>Do you guys have any plan to patch this change to libevent2.0.x ?

2.0.x has a lot of other issues, and it is already kind of EOL, but I will take a look if you will provide normal reproducer.

>moving to latest libevent 2.1.x is not possible now

Just out of curiosity what are the issues do not allows you to use 2.1.x?

**Comment by PugalendranDell** [comment]
@azat - This fix is available only in latest libevent 2.1.11 which is not a "release candidate"  right ? 

I am working on the reproducer. will get back soon

**Comment by PugalendranDell** [comment]
I am okay to use libevent 2.1.7. But this fix is not available.

**Comment by azat** [comment]
>This fix is available only in latest libevent 2.1.11 which is not a "release candidate" right ?

The patch in upstream is  0b4b0ef.

```
$ git tag | grep ^release-2.1. | xargs -n1 -i -r sh -c 'git log --grep 0b4b0ef -1 {} | fgrep -q 0b4b0ef && echo {}'
release-2.1.11-stable
release-2.1.10-stable
release-2.1.9-beta
```

>I am okay to use libevent 2.1.7. But this fix is not available.

Why you are using 2.1.7-rc instead of 2.1.11-stable?

**Comment by PugalendranDell** [comment]
@azat - Thought only release candidate is advisable to use, will use 2.1.11-stable version. Thanks
